packages/cli/src/handlers/dbt/compile.ts

```diff
@@ -197,4 +157,73 @@ export const dbtList = async (
             `Error executing 'dbt ls':\n  ${msg}\nEnsure you're on the latest patch version. '--use-dbt-list' is true by default; if you encounter issues, try using '--use-dbt-list=false`,
         );
     }
-};
+}
+
+export async function maybeCompileModelsAndJoins(
+    loadManifestOpts: LoadManifestArgs,
+    options: DbtCompileOptions,
+): Promise<string[] | undefined> {
+    // Skipping assumes manifest.json already exists.
+    if (options.skipDbtCompile) {
+        GlobalState.debug('> Skipping dbt compile');
+        return undefined;
+    }
+
+    // do initial compilation so we can get the list of models that are compiled after this command (e.g. selecting/excluding by tags)
+    let compiledModelIds: string[] | undefined;
+    if (options.useDbtList) {
+        compiledModelIds = await dbtList(options);
+    } else {
+        await dbtCompile(options);
+    }
+
+    // If no models are explicitly selected or excluded, we don't need to explicitly find joined models
+    if (!options.select && !options.exclude) {
+        return compiledModelIds;
+    }
+
+    // Load manifest and get all models
+    const manifest = await loadManifest(loadManifestOpts);
+    const allManifestModels = getModelsFromManifest(manifest);
+    const currCompiledModels = getCompiledModels(
+        allManifestModels,
+        compiledModelIds,
+    );
+
+    // Get models and their joined models
+    const requiredModels = new Set(
+        currCompiledModels.reduce<string[]>((acc, model) => {
+            const joinedModelNames = getJoinedModelsRecursively(
+                model,
+                allManifestModels,
+                new Set(acc), // minimize recursion by passing already visited models in the current list
+            );
+            return [...acc, model.name, ...joinedModelNames];
+        }, []),
+    );
+
+    const requiredModelsNames = Array.from(requiredModels);
+    const missingJoinedModels = xor(
+        requiredModelsNames,
+        currCompiledModels.map((model) => model.name),
+    );
+    if (missingJoinedModels.length > 0) {
```

Comment:
Avoid compiling the second time if all the models are present. 
Command to test locally:
```
node ./packages/cli/dist/index.js preview --project-dir ./examples/full-jaffle-shop-demo/dbt --profiles-dir ./examples/full-jaffle-shop-demo/profiles -s customers plan membership
```